sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala

```diff
@@ -79,11 +79,18 @@ case class LogicalRelation(
   /** Used to lookup original attribute capitalization */
   val attributeMap: AttributeMap[AttributeReference] = AttributeMap(output.map(o => (o, o)))
 
-  def newInstance(): this.type =
+  /**
+   * Returns a new instance of this LogicalRelation. According to the semantics of
+   * MultiInstanceRelation, this method should returns a copy of this object with
+   * unique expression ids. Thus we don't respect the `expectedOutputAttributes` and
+   * create new instances of original output attributes.
+   */
+  override def newInstance(): this.type = {
     LogicalRelation(
       relation,
-      expectedOutputAttributes,
```

Comment:
if `expectedOutputAttributes` is None, `output` is defined as `relation.schema.toAttributes`, we will get different expr ids.